loguru.hpp

```diff
@@ -1379,29 +1392,100 @@ namespace loguru
+#endif
 	void file_log(void* user_data, const Message& message)
 	{
-		FILE* file = reinterpret_cast<FILE*>(user_data);
+#ifdef LOGURU_WITH_FILEABS
+		pthread_mutex_lock(&(reinterpret_cast<FileAbs*>(user_data)->lock));
+		file_reopen(user_data);
+		FILE* file = to_file(user_data);
+		if(!file) {
+			pthread_mutex_unlock(&(reinterpret_cast<FileAbs*>(user_data)->lock));
+			return;
+		}
+#else
+		FILE* file = to_file(user_data);
+#endif
 		fprintf(file, "%s%s%s%s\n",
 			message.preamble, message.indentation, message.prefix, message.message);
 		if (g_flush_interval_ms == 0) {
 			fflush(file);
 		}
+#ifdef LOGURU_WITH_FILEABS
+		pthread_mutex_unlock(&(reinterpret_cast<FileAbs*>(user_data)->lock));
+#endif
 	}
 
 	void file_close(void* user_data)
 	{
-		FILE* file = reinterpret_cast<FILE*>(user_data);
-		fclose(file);
+		FILE* file = to_file(user_data);
+		if(file) {
+			fclose(file);
+		}
+#ifdef LOGURU_WITH_FILEABS
+		delete reinterpret_cast<FileAbs*>(user_data);
+#endif
 	}
 
 	void file_flush(void* user_data)
 	{
-		FILE* file = reinterpret_cast<FILE*>(user_data);
+		FILE* file = to_file(user_data);
 		fflush(file);
 	}
 
+#ifdef LOGURU_WITH_FILEABS
+	void add_callback_abs(const char* id, log_handler_t callback, void* user_data,
+						  Verbosity verbosity, close_handler_t on_close, flush_handler_t on_flush)
+	{
+		std::lock_guard<std::recursive_mutex> lock(s_mutex);
+		s_callbacks.push_back(Callback{id, callback, user_data, verbosity, on_close, on_flush, 0});
+	}
+
+	bool remove_callback_abs(const char* id)
+	{
+		std::lock_guard<std::recursive_mutex> lock(s_mutex);
+		auto it = std::find_if(begin(s_callbacks), end(s_callbacks), [&](const Callback& c) { return c.id == id; });
+		if (it != s_callbacks.end()) {
+			s_callbacks.erase(it);
+			return true;
+		} else {
+			LOG_F(ERROR, "Failed to locate callback with id '%s'", id);
+			return false;
+		}
+	}
+
+	void file_reopen(void* user_data)
+	{
+		FileAbs * file_abs = reinterpret_cast<FileAbs*>(user_data);
+		struct stat st;
+		int ret;
+		if (!file_abs->fp || (ret = stat(file_abs->path, &st)) == -1 || (st.st_ino != file_abs->st.st_ino)) {
+			remove_callback_abs(file_abs->path); //adapted version, only remove from callback vector s_callbacks;
+			if (ret < 0) {
+				LOG_F(INFO, "Reopening file '%s' due to %m", file_abs->path); // still logging to previous fd;
+			} else {
+				LOG_F(INFO, "Reopening file '%s' due to file changed", file_abs->path);
+			}
+			// try reopen current file.
+			if (!create_directories(file_abs->path)) {
+				LOG_F(ERROR, "Failed to create directories to '%s'", file_abs->path);
+			}
+			file_abs->fp = fopen(file_abs->path, file_abs->mode_str);
+			if (!file_abs->fp) {
+				LOG_F(ERROR, "Failed to open '%s'", file_abs->path);
+			} else {
+				stat(file_abs->path, &file_abs->st);
+			}
+			// adapted version, only add to callback vector s_callbacks;
+			add_callback_abs(file_abs->path, file_log, file_abs, file_abs->verbosity, file_close, file_flush);
```

Comment:
Can we get rid of the `remove_callback_abs` and `add_callback_abs` somehow? They contribute a lot of code and complexity, and reorders the order of the logging callbacks, which can be surprising.

I'm assuming they are only there so you can log things during file reopening, without infinite recursion? If so, I feel something much more simpler would be to have a single `bool is_reopening = false;` in `FileAbs` which is set to `true` during the logic above. Thus we could check `is_reopening` in `file_reopen` to avoid infinite recursion.